net.bobbo.player-controller/player_scripts/crouch.gd

```diff
@@ -33,4 +40,37 @@ func player_physics_process(delta: float):
 	var scale_difference = new_scale - previous_scale
 	
 	player.velocity.y += scale_difference * 0.5
-	player.height = new_scale
+	
+	# Make it so that the crouch can only go as low as regularly crouching and only high enough such that the player doesn't get stuck if an object is detected above them
+	player.height = clamp(new_scale, _max_player_height / 2, _height_above)
+	
+	# Get the 3D physics space from a Node3D like the player_pivot
```

Comment:
The whole raycasting section could likely be cleaned up by refactoring it into some private functions. The body of `player_physics_process` is becoming a little cluttered for my tastes!

Reply:
In combination with previous notes - I suggest perhaps making a `_should_force_crouch(): bool` function that returns true if the player should be forced to crouch (such as if there is a ceiling above them still) and false otherwise.

Then at the start of `player_physics_process` you can simply call something like:
```
if Input.is_action_pressed(crouch_action) or _should_force_crouch():
	crouch_target = 0.5
else:
	crouch_target = 1
```

Reply:
That's a pretty good idea that would definitely clean up the code a bit, definitely gonna go with something like that.